klippy/mcu.py

```diff
@@ -230,27 +230,31 @@ def _build_config(self):
             return
         self._oid = self._mcu.create_oid()
         self._mcu.add_config_cmd(
-            "config_soft_pwm_out oid=%d pin=%s cycle_ticks=%d value=%d"
+            "config_soft_pwm_out oid=%d pin=%s value=%d"
             " default_value=%d max_duration=%d"
-            % (self._oid, self._pin, cycle_ticks,
-               self._start_value >= 1.0, self._shutdown_value >= 0.5,
-               self._mcu.seconds_to_clock(self._max_duration)))
+            % (self._oid, self._pin, self._start_value >= 1.0,
+                self._shutdown_value >= 0.5,
+                self._mcu.seconds_to_clock(self._max_duration)))
         curtime = self._mcu.get_printer().get_reactor().monotonic()
         printtime = self._mcu.estimated_print_time(curtime)
         self._last_clock = self._mcu.print_time_to_clock(printtime + 0.100)
         svalue = int(self._start_value * self._pwm_max + 0.5)
         self._mcu.add_config_cmd(
-            "schedule_soft_pwm_out oid=%d clock=%d on_ticks=%d"
-            % (self._oid, self._last_clock, svalue), is_init=True)
+            "schedule_soft_pwm_out oid=%d clock=%d on_ticks=%d off_ticks=%d"
+            % (self._oid, self._last_clock, svalue, cycle_ticks - svalue),
+                is_init=True)
         self._set_cmd = self._mcu.lookup_command(
-            "schedule_soft_pwm_out oid=%c clock=%u on_ticks=%u",
+            "schedule_soft_pwm_out oid=%c clock=%u on_ticks=%u off_ticks=%u",
             cq=cmd_queue)
-    def set_pwm(self, print_time, value):
+    def set_pwm(self, print_time, value, cycle_time=None):
+        if cycle_time == None:
+            cycle_time=self._cycle_time
+        cycle_ticks = self._mcu.seconds_to_clock(cycle_time)
         clock = self._mcu.print_time_to_clock(print_time)
         if self._invert:
             value = 1. - value
-        value = int(max(0., min(1., value)) * self._pwm_max + 0.5)
-        self._set_cmd.send([self._oid, clock, value],
+        on_ticks = int(max(0., min(1., value)) * float(cycle_ticks) + 0.5)
+        self._set_cmd.send([self._oid, clock, on_ticks, cycle_ticks - on_ticks],
```

Comment:
It looks like this change would break the hard_pwm code.

Reply:
Is it because we send a fourth parameter to the `schedule_pwm_out` function when it expects 3? Then maybe add a fourth parameter to that function and leave it unused for now?

Reply:
On a second thought, wouldn't such a mistake fail the automated test?

Reply:
Yes.  I didn't check if it would actually break hard_pwm, but in the hard pwm case there is: `self._set_cmd = self._mcu.lookup_command(
                "schedule_pwm_out oid=%c clock=%u value=%hu", cq=cmd_queue)`
which would not be compatible with the new soft pwm case.

It should be possible to check for `self._hardware_pwm` in `set_pwm()` and issue the appropriate command.

It should be possible to expand the test cases to do more hard pwm tests.  (Since the above looks like a regression, and the test cases didn't fail, it looks like this isn't being tested.)

Reply:
I looked into the tests and found no SET_PIN commands anywhere. So I wrote soft and hard pwm tests. I had to rewrite the data dictionaries because they contain the previous declarations of `schedule_soft_pwm_out` and `config_soft_pwm_out`.
The hard pwm test didn't fail. Should I check for `self._hardware_pwm` anyway?

Reply:
Ah, `send_cmd()` calls `msgproto:MessageFormat:encode()` which does not raise an error on too many parameters.  In any case, yes, I think the code would be more clear if there was a separate branch for `self._hardware_pwm`.

Reply:
Ok, I'll add the check.